### PR TITLE
Run2-alca96 Add propagation tool from Hcal to Ecal

### PR DESCRIPTION
**Pull request by bsunanda**
Also add some additional tests for single particle response studies

### CODE REVIEW

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by arunhep** [comment]
+1

File: Calibration/IsolatedParticles/src/CaloPropagateTrack.cc [hunk 1]
```diff
@@ -601,6 +602,27 @@ namespace spr{
     return spr::propagateECAL (vertex, momentum, charge, bfield, debug);
   }
 
+  std::pair<DetId,bool> propagateIdECAL(const HcalDetId& id, const CaloGeometry* geo, const MagneticField* bField, bool debug) {
+
+    const HcalGeometry* gHB = (const HcalGeometry*)(geo->getSubdetectorGeometry(DetId::Hcal,HcalBarrel));
+    GlobalPoint vertex(0,0,0);
```
Comment (davidlange6) [inline Calibration/IsolatedParticles/src/CaloPropagateTrack.cc]: hi @bsunanda - could this be declared const?

File: Calibration/IsolatedParticles/src/CaloPropagateTrack.cc [hunk 2]
```diff
@@ -601,6 +602,27 @@ namespace spr{
...
+    GlobalPoint hit(gHB->getPosition(id));
+    GlobalVector momentum = GlobalVector(hit.x(),hit.y(),hit.z());
+    std::pair<math::XYZPoint,bool> info = propagateECAL(vertex,momentum,0,bField,debug);
+    DetId eId(0);
+    if (info.second) {
+      const GlobalPoint point(info.first.x(),info.first.y(),info.first.z());
+      if (std::abs(point.eta())<spr::etaBEEcal) {
+	const EcalBarrelGeometry *barrelGeom = (dynamic_cast< const EcalBarrelGeometry *> (geo->getSubdetectorGeometry(DetId::Ecal,EcalBarrel)));
```
Comment (davidlange6) [inline Calibration/IsolatedParticles/src/CaloPropagateTrack.cc]: I don't think this cast is needed. getClosestCell is part of the base class.

**Comment by davidlange6** [comment]
-1
pending answers to review comments

File: Calibration/IsolatedParticles/src/CaloPropagateTrack.cc (hunk 2 above)
Comment (bsunanda) [reply Calibration/IsolatedParticles/src/CaloPropagateTrack.cc]: This is overridden in the EcalBarrelGeometry.h, EcalEndcapGeometry.h, HcalGeometry.H

Comment (bsunanda) [reply Calibration/IsolatedParticles/src/CaloPropagateTrack.cc]: There is no way I can avoid the casting here

File: Calibration/IsolatedParticles/src/CaloPropagateTrack.cc (hunk 1 above)
Comment (bsunanda) [reply Calibration/IsolatedParticles/src/CaloPropagateTrack.cc]: Added const to all GlobalPoint, GlobalVector where applicable

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by arunhep** [comment]
+1

File: Calibration/IsolatedParticles/src/CaloPropagateTrack.cc (hunk 2 above)
Comment (davidlange6) [reply Calibration/IsolatedParticles/src/CaloPropagateTrack.cc]: hi @bsunanda -i'm not sure I follow - your code does not need to know the type - as of course the pointer will know the type and call the appropriate function (unless something strange has been done upstream?)

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@davidlange6 Now I corrected this

**Comment by bsunanda** [comment]
@arunhep @davidlange6 Please approve and accept this PR

**Comment by arunhep** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Pull request merged**